sysfile.c

```diff
@@ -465,3 +465,20 @@ sys_forkexec(void)
   }
   return forkexec(path, argv);
 }
+
+int
+sys_shmem_alloc(void)
+{
+  return (int) shmem_alloc();
```

Comment:
I'm taking this from `sbrk`, which is defined to return `int` for `sys_sbrk`, but actually returns `char*`.

Reply:
for most of the syscalls the return value designates the status of the operation. here you need to return a pointer. sbrk is returning -1 in case of error. that is why they are using int. anyway int is not a good type to hold address as it has a sign bit.

Reply:
I don't get your point. `sys_sbrk` is still returning the address as an `int` while `sz` is of type `uint`.

Reply:
i don't agree with sbrk implementation either. But it is kind of okay since sz is not an address. it denotes the size of the process. but in your case you are returning an actual address

Reply:
the return type is chosen to be large enough to hold a pointer value any way. see https://stackoverflow.com/questions/20940212/why-is-linux-syscall-return-type-long

Reply:
precisely, int is not large enough to hold 32-bit address. we don't need to use long since we are not writing 64-bit kernel.

Reply:
`long` on 32 bit arch is 32 bits long.

I'll stick with this API for consistency with `sbrk` and `mmap` which I'll implement later.

Reply:
First of all, size of long depends on the compiler. the only restriction imposed by POSIX is short<=int<=long. long can be 64-bit as well which usually is I think. Also shmem is more similar to malloc than sbrk. Check the linux implementation of brk/sbrk here. https://linux.die.net/man/2/sbrk.
As being said, it is not a must have change since you are only allocating lower half of address space.

Reply:
Great, we're in agreement, then.

However, there are a few points to make clear.
- `brk` is implemented at https://github.com/torvalds/linux/blob/2db767d9889cef087149a5eaa35c1497671fa40f/mm/mmap.c#L182
- `DEFINE_SYSCALL1` is defined at https://github.com/torvalds/linux/blob/2db767d9889cef087149a5eaa35c1497671fa40f/include/linux/syscalls.h#L196. Note the return value is `asmlinkage long sys##name(__MAP(x,__SC_DECL,__VA_ARGS__))`.
- That means that `brk` actually returns a `long` but the signature for the syscall just throws away the higher bits (possibly, depending on the architecture) by returning an `int`.
- However, `sbrk`, as implemented in GNU glibc cheats and recovers the information in the higher bits as defined at code.metager.de/source/xref/gnu/glibc/misc/sbrk.c and http://code.metager.de/source/xref/gnu/glibc/sysdeps/unix/sysv/linux/x86_64/brk.c